File: src/main/java/sigmaone/industrialism/block/multiblock/BlockEntityMultiblockRootBase.java
package sigmaone.industrialism.block.multiblock;

import net.minecraft.block.Block;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityType;

import java.util.HashSet;

public class BlockEntityMultiblockRootBase extends BlockEntity {
    protected HashSet<BlockEntityMultiblockChildBase> children;
    protected Block[][][] layout;
    protected int[] posInLayout;

    public BlockEntityMultiblockRootBase(BlockEntityType<?> type) {
        super(type);
        this.children = new HashSet<>();
    }

    public void setLayout(Block[][][] layout) {
        this.layout = layout;
    }

    public void setPosInLayout(int[] pos) {
        this.posInLayout = pos;
    }

    public void addChild(BlockEntityMultiblockChildBase child) {
        this.children.add(child);
    }

    public void disassemble() {
        for (BlockEntityMultiblockChildBase child : this.children) {
            if (this.getWorld() != null && this.getWorld().getBlockState(child.getPos()) != null) {
                this.getWorld().setBlockState(child.getPos(), (this.layout[child.getPosInLayout()[1]][child.getPosInLayout()[0]][child.getPosInLayout()[2]].getDefaultState()));
                this.getWorld().removeBlockEntity(child.getPos());
            }
        }
        if (this.getWorld() != null) {
            this.getWorld().setBlockState(this.getPos(), this.layout[this.posInLayout[1]][this.posInLayout[0]][this.posInLayout[2]].getDefaultState());
            this.getWorld().removeBlockEntity(this.getPos());
        }
    }
}

File: src/main/kotlin/sigmaone/industrialism/block/multiblock/BlockEntityMultiblockRootBase.kt
package sigmaone.industrialism.block.multiblock

import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtHelper
import net.minecraft.util.math.BlockPos
import kotlin.collections.ArrayList

abstract class BlockEntityMultiblockRootBase(type: BlockEntityType<*>?) : BlockEntity(type) {
    var children: ArrayList<BlockPos> = ArrayList()

    fun addChild(pos: BlockPos) {
        children.add(pos)
        markDirty()
    }

    fun disassemble() {
        for (child in children) {
            world!!.setBlockState(child, Blocks.AIR.defaultState)
        }
        world!!.setBlockState(pos, Blocks.AIR.defaultState)
    }

    override fun fromTag(state: BlockState?, tag: CompoundTag?) {
        super.fromTag(state, tag)
        val childrenTag: CompoundTag = tag!!.getCompound("children")
        for (i in 0..childrenTag.getInt("amount")) {
            children.add(NbtHelper.toBlockPos(childrenTag.getCompound(i.toString())))
        }
    }

    override fun toTag(tag: CompoundTag?): CompoundTag {
        super.toTag(tag)
        val childrenTag = CompoundTag()
        for ((i, child) in children.withIndex()) {
            childrenTag.put(i.toString(), NbtHelper.fromBlockPos(child))
        }
        childrenTag.putInt("amount", children.size)
        tag!!.put("children", childrenTag)
        return tag
    }
}